src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala

```diff
@@ -93,6 +94,58 @@ final case class PurposeApiServiceImpl(
     }
   }
 
+  override def createPurposeFromEService(eServiceId: String, seed: ProducerPurposeSeed)(implicit
+    contexts: Seq[(String, String)],
+    toEntityMarshallerPurpose: ToEntityMarshaller[Purpose],
+    toEntityMarshallerProblem: ToEntityMarshaller[Problem]
+  ): Route = authorize(ADMIN_ROLE) {
+    val operationLabel =
+      s"Creating Purposes for EService ${eServiceId}, Consumer ${seed.consumerId}"
+
+    val result: Future[Purpose] = for {
+      organizationId <- getOrganizationIdFutureUUID(contexts)
+      _              <- assertOrganizationIsAConsumer(organizationId, seed.consumerId)
+      eServiceUUID   <- eServiceId.toFutureUUID
+      eService       <- catalogManagementService.getEServiceById(eServiceUUID)
+      _ <- if (eService.mode == Deliver) Future.failed(EServiceNotInReceiveMode(eService.id)) else Future.unit
+      riskAnalysis <- eService.riskAnalysis
+        .filter(_.id == seed.riskAnalysisId)
+        .headOption
+        .toFuture(RiskAnalysisNotFound(eServiceUUID, seed.riskAnalysisId))
+      _            <-
+        if (seed.isFreeOfCharge && seed.freeOfChargeReason.isEmpty) Future.failed(MissingFreeOfChargeReason)
+        else Future.unit
+      tenant       <- tenantManagementService.getTenantById(organizationId)
+      tenantKind   <- tenant.kind.toFuture(TenantKindNotFound(tenant.id))
+      purposeSeed = seed.toManagement(eServiceUUID, riskAnalysis.riskAnalysisForm.toManagement(seed.riskAnalysisId))
+      agreements <- agreementManagementService.getAgreements(eServiceUUID, seed.consumerId, OPERATIVE_AGREEMENT_STATES)
+      agreement  <- agreements.headOption.toFuture(AgreementNotFound(eServiceUUID.toString, seed.consumerId.toString))
+      maybePurpose <- purposeManagementService
+        .listPurposes(
+          seed.consumerId,
+          seed.title.some,
+          List(agreement.eserviceId.toString),
+          List(agreement.consumerId.toString),
+          List(agreement.producerId.toString),
+          states = List.empty,
+          excludeDraft = false,
+          offset = 0,
+          limit = 1,
+          exactMatchOnTitle = true
+        )
+        .map(_.results.headOption)
+
+      _       <- maybePurpose.fold(Future.unit)(_ => Future.failed(DuplicatedPurposeName(seed.title)))
+      purpose <- purposeManagementService.createPurpose(purposeSeed)
+      isValidRiskAnalysisForm = isRiskAnalysisFormValid(
+        riskAnalysisForm = purpose.riskAnalysisForm.map(_.toApi),
+        schemaOnlyValidation = true
+      )(tenantKind)
```

Comment:
to avoid code duplication I suggest moving all these commons steps in a function, and then call it both from `createPurposeFromEService` and `createPurpose`

Reply:
Taaaccccc!
Seed type is different. It is not the same code. However I will separate code in multiple functions.

Reply:
seed "type" is different or just how it is built? I guess the type is the same considering it is passed to the same function.
in the latter I would suggest to just pass the seed as parameter to the function

Reply:
the latest pushed solution works for me 👍